File: src/neetcode/Java/_37SudokuSolver.java
package neetcode.Java;

public class _37SudokuSolver {
    public void solveSudoku(char[][] board) {
        sodoku(board);
    }

    boolean sodoku(char board[][]) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    for (char c = '1'; c <= '9'; c++) {
                        if (isValid(board, i, j, c)) {
                            board[i][j] = c;
                            if (sodoku(board))
                                return true;
                            else
                                board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    boolean isValid(char[][] board, int row, int col, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == c)
                return false;
            if (board[row][i] == c)
                return false;
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c)
                return false;
        }
        return true;
    }

    public static void main(String[] args) {
        _37SudokuSolver obj = new _37SudokuSolver();
        {
            char[][] board = {{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                    {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                    {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                    {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                    {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                    {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                    {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                    {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                    {'.', '.', '.', '.', '8', '.', '.', '7', '9'}};

            obj.solveSudoku(board);

            for (char[] i : board) {
                for (char n : i) {
                    System.out.print(n + "   ");
                }
                System.out.println();
            }
        }
    }
}

File: src/neetcode/kotlin/_37SudokuSolver.kt
package neetcode.kotlin

class _37SudokuSolver {
    fun solveSudoku(board: Array<CharArray>) {
        sodoku(board)
    }

    fun sodoku(board: Array<CharArray>): Boolean {
        for (i in 0..8) {
            for (j in 0..8) {
                if (board[i][j] == '.') {
                    var c = '1'
                    while (c <= '9') {
                        if (isValid(board, i, j, c)) {
                            board[i][j] = c
                            if (sodoku(board)) return true
                            else board[i][j] = '.'
                        }
                        c++
                    }
                    return false
                }
            }
        }
        return true
    }

    fun isValid(board: Array<CharArray>, row: Int, col: Int, c: Char): Boolean {
        for (i in 0..8) {
            if (board[i][col] == c) return false
            if (board[row][i] == c) return false
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c) return false
        }
        return true
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _37SudokuSolver()
            run {
                val board = arrayOf(
                    charArrayOf('5', '3', '.', '.', '7', '.', '.', '.', '.'),
                    charArrayOf('6', '.', '.', '1', '9', '5', '.', '.', '.'),
                    charArrayOf('.', '9', '8', '.', '.', '.', '.', '6', '.'),
                    charArrayOf('8', '.', '.', '.', '6', '.', '.', '.', '3'),
                    charArrayOf('4', '.', '.', '8', '.', '3', '.', '.', '1'),
                    charArrayOf('7', '.', '.', '.', '2', '.', '.', '.', '6'),
                    charArrayOf('.', '6', '.', '.', '.', '.', '2', '8', '.'),
                    charArrayOf('.', '.', '.', '4', '1', '9', '.', '.', '5'),
                    charArrayOf('.', '.', '.', '.', '8', '.', '.', '7', '9')
                )
                obj.solveSudoku(board)
                for (i in board) {
                    for (n in i) {
                        print("$n   ")
                    }
                    println()
                }
            }
        }
    }
}
